source/server/connection_handler_impl.cc

```diff
@@ -83,16 +81,27 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
   if (auto* listener = details.listener_->listener(); listener != nullptr) {
     listener->setRejectFraction(listener_reject_fraction_);
   }
-  listeners_.emplace_back(config.listenSocketFactory().localAddress(), std::move(details));
+
+  ASSERT(!listener_map_by_tag_.contains(config.listenerTag()));
+
+  listener_map_by_tag_.emplace(
+      config.listenerTag(),
+      std::make_pair(config.listenSocketFactory().localAddress(), std::move(details)));
+  // This map only store the new listener.
+  listener_map_by_address_.insert_or_assign(config.listenSocketFactory().localAddress()->asString(),
+                                            &listener_map_by_tag_[config.listenerTag()].second);
 }
 
 void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
-  for (auto listener = listeners_.begin(); listener != listeners_.end();) {
-    if (listener->second.listener_->listenerTag() == listener_tag) {
-      listener = listeners_.erase(listener);
-    } else {
-      ++listener;
+  auto listener_iter = listener_map_by_tag_.find(listener_tag);
+  if (listener_iter != listener_map_by_tag_.end()) {
+    // listener_map_by_address_ may already update to the new listener. Compare it with the one
+    // which find from listener_map_by_tag_, only delete it when it is same listener.
+    if (listener_map_by_address_[listener_iter->second.first->asString()] ==
+        &listener_iter->second.second) {
+      listener_map_by_address_.erase(listener_iter->second.first->asString());
```

Comment:
IMO, it is not a good idea to use addresses to distinguish between old and new values. Because the memory may be reallocated. 
May be we should store the related listener tag the the `listener_map_by_address_`.

Reply:
got it.